Mathlib/Data/QuotLike.lean

```diff
@@ -0,0 +1,666 @@
+  QuotLike.ind (QuotLike.ind h qa) qb
+
+@[elab_as_elim]
+protected theorem inductionOn₂ {motive : Qa → Qb → Prop}
+    (qa : Qa) (qb : Qb)
+    (h : (a : α) → (b : β) → motive ⟦a⟧ ⟦b⟧) :
+    motive qa qb :=
+  QuotLike.ind₂ h qa qb
+
+/-- A binary version of `Quot.recOnSubsingleton`. -/
+@[elab_as_elim]
+protected def recOnSubsingleton₂ {motive : Qa → Qb → Sort*}
+    [_h : ∀ a b, Subsingleton (motive ⟦a⟧ ⟦b⟧)]
+    (qa : Qa) (qb : Qb) (f : ∀ a b, motive ⟦a⟧ ⟦b⟧) :
+    motive qa qb :=
+  QuotLike.recOnSubsingleton (_h := fun _ ↦ QuotLike.ind inferInstance qb) qa
+    fun a ↦ QuotLike.recOnSubsingleton qb fun b ↦ f a b
+
+/-- Recursion on two `QuotLike` arguments `qa` and `qb`, result type depends on `⟦a⟧` and `⟦b⟧`. -/
+@[elab_as_elim]
+protected def hrecOn₂ [IsRefl α ra] [IsRefl β rb] {motive : Qa → Qb → Sort*}
+    (qa : Qa) (qb : Qb) (f : ∀ a b, motive ⟦a⟧ ⟦b⟧)
+    (h : ∀ a₁ b₁ a₂ b₂, ra a₁ a₂ → rb b₁ b₂ → HEq (f a₁ b₁) (f a₂ b₂)) :
+    motive qa qb :=
+  QuotLike.hrecOn qa
+    (fun a ↦ QuotLike.hrecOn qb (f a) (fun b₁ b₂ pb ↦ h _ _ _ _ (refl _) pb))
+    fun a₁ a₂ pa ↦ by exact QuotLike.inductionOn qb fun b ↦ by simp [h, pa, refl]
+
+@[simp]
+theorem hrecOn₂_mkQ [IsRefl α ra] [IsRefl β rb] {motive : Qa → Qb → Sort*}
+    (a : α) (b : β) (f : ∀ a b, motive ⟦a⟧ ⟦b⟧)
+    (h : ∀ a₁ b₁ a₂ b₂, ra a₁ a₂ → rb b₁ b₂ → HEq (f a₁ b₁) (f a₂ b₂)) :
+    QuotLike.hrecOn₂ (motive := motive) ⟦a⟧ ⟦b⟧ f h = f a b := by
+  simp [QuotLike.hrecOn₂]
+
+@[elab_as_elim]
+protected theorem ind₃ {motive : Qa → Qb → Qc → Prop}
+    (h : (a : α) → (b : β) → (c : γ) → motive ⟦a⟧ ⟦b⟧ ⟦c⟧)
+    (qa : Qa) (qb : Qb) (qc : Qc) :
+    motive qa qb qc :=
+  QuotLike.ind (QuotLike.ind₂ h qa qb) qc
+
+@[elab_as_elim]
+protected theorem inductionOn₃ {motive : Qa → Qb → Qc → Prop}
+    (qa : Qa) (qb : Qb) (qc : Qc)
+    (h : (a : α) → (b : β) → (c : γ) → motive ⟦a⟧ ⟦b⟧ ⟦c⟧) :
+    motive qa qb qc :=
+  QuotLike.ind₃ h qa qb qc
+
+end
+
+section
+
+variable {Q α : Sort*} {r : α → α → Prop} [QuotLike Q α r]
+
+/-- Makes a quotient from `Quot r`. -/
+def ofQuot : Quot r → Q :=
+  Quot.lift mkQ fun _ _ ↦ sound
+
+@[simp]
+theorem ofQuot_toQuot (q : Q) : ofQuot (toQuot q) = q :=
+  ind (fun a ↦ by rw [toQuot_mkQ, ofQuot]) q
+
+@[simp]
+theorem toQuot_ofQuot (q : Quot r) : toQuot (ofQuot q : Q) = q :=
+  Quot.ind (β := fun q ↦ toQuot (ofQuot q) = q) toQuot_mkQ q
+
+theorem toQuot_injective : Function.Injective (toQuot (Q := Q)) :=
+  Function.LeftInverse.injective ofQuot_toQuot
+
+theorem eq_iff_quotMk {a b : α} : (⟦a⟧ : Q) = ⟦b⟧ ↔ Quot.mk r a = Quot.mk r b := by
+  rw [← toQuot_mkQ (Q := Q), ← toQuot_mkQ (Q := Q), toQuot_injective.eq_iff]
+
+theorem Quot.exact {α r} [IsEquiv α r] {a b : α} : Quot.mk r a = Quot.mk r b → r a b :=
+  Quotient.exact (s := ⟨r, refl, symm, _root_.trans⟩)
+
+theorem exact [IsEquiv α r] {a b : α} : (⟦a⟧ : Q) = ⟦b⟧ → r a b :=
+  fun h ↦ Quot.exact (eq_iff_quotMk.mp h)
+
+@[simp]
+theorem eq [IsEquiv α r] {a b : α} : (⟦a⟧ : Q) = ⟦b⟧ ↔ r a b :=
```

Comment:
I don't believe this should be a `simp` theorem. The reason we often work with quotients is because equality is nicer to reason about than an arbitrary equivalence relation; this theorem only helps to undo this.

I don't think `Quotient.eq` should be `simp` either, but that's a question for another time.

Reply:
But at the same time, this is about any equality in the quotient, this is about equality of two values of the quotient map. This lemma looks like a good simp lemma to me